Encryption.md

```diff
@@ -19,216 +19,544 @@
+   Master key using one of the options described above – but the resulting key material is stored 
+   separately, outside the data file, and will be retrieved using the counter and file path.
+   * Any of the three above, plus a string ID of the KMS instance (in case there are many).
+   * Random string - useful for creation of footer keys in organizations not willing to authenticate 
+   and connect all readers to a KMS. The footer key is generated as F(key_metadata), where F is an 
+   algorithm unknown outside the organization. These keys are less secure than KMS-managed keys, but 
+   better than using no keys at all in plaintext and unsigned footers (discussed in the sections 5.2 
+   and 5.3.)
+   
+Key metadata can also be empty - in a case the encryption keys are fully managed by the caller 
+code, and passed explicitly to Parquet for the footer and each encrypted column.
+
+### 4.4 Additional Authenticated Data
+The AES GCM cipher in the basic mode protects against byte replacement inside a ciphertext, 
+but can't prevent replacement of one ciphertext with another (encrypted with the same key). 
+This can be solved by the AAD mode, using different AADs for different ciphertexts. Parquet 
+modular encryption uses the AAD mode to protect against swapping ciphertext modules inside 
+a file, between files - or against swapping full files (for example, replacement of a file 
+with an old version). Obviously, the AAD must reflect the identity of a file and of the modules 
+inside the file.
+
+Parquet constructs a module AAD from two components: an optional AAD_Prefix - a string provided 
+by the user for the file, and an AAD_Suffix, built internally for each GCM-encrypted module 
+inside the file. The AAD prefix should reflect the target identity that helps to detect file 
+swapping (simple example - table name with a date and partition, e.g., "employees_23_05_2018.part0"). 
+The AAD suffix reflects the internal identity of modules inside the file, which for example 
+prevents replacement of column pages in row group 0 by pages from the same column in row 
+group 1 (when columns are encrypted with the CGM cipher). The module AAD is a direct concatenation 
+of the prefix and suffix parts.
+
+#### 4.4.1 AAD_Prefix
+A file writer passes an optional AAD_Prefix string upon file creation, that allows to differentiate 
+it from other files. The reader should know this string apriori, or should be able to 
+retrieve and verify it, using a convention accepted in the organization. An optional `aad_metadata` 
+byte array can be specified in a file - to help a reader retrieve the AAD 
+prefix string used for file enciphering. The aad_metadata  field can be empty, since readers might 
+know apriori a right AAD_Prefix for each file. Alternatively, the aad_metadata  can contain hints 
+that help readers to get the right AAD prefix. Yet another option is to keep the AAD_Prefix itself 
```

Comment:
This sentence, "aad_metadata can contain hints that help readers to get the right AAD prefix" is a duplicate of "to help a reader retrieve the AAD prefix string used for file enciphering" above.